### PR TITLE
Add array_intersect UDF to Presto

### PR DESCRIPTION
**Pull request by ameydhar**

### CODE REVIEW

File: presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java [hunk 1]
```diff
@@ -282,6 +282,20 @@ public void testSort()
...
     @Test
+    public void testArrayIntersect()
+            throws Exception
+    {
+        assertFunction("ARRAY_INTERSECT(ARRAY [12], ARRAY [10])", Lists.newArrayList());
+        assertFunction("ARRAY_INTERSECT(ARRAY['foo', 'bar', 'baz'] , ARRAY['foo', 'test', 'bar'])", ImmutableList.of("foo", "bar"));
+        assertFunction("ARRAY_INTERSECT(ARRAY [1, 5], ARRAY [1])", ImmutableList.of(1L));
+        assertFunction("ARRAY_INTERSECT(ARRAY [1, 1, 2, 4], ARRAY [1, 3, 4, 4])", ImmutableList.of(1L, 1L, 4L));
+        assertFunction("ARRAY_INTERSECT(ARRAY [2, 8], ARRAY [8, 3])", ImmutableList.of(8L));
+        assertFunction("ARRAY_INTERSECT(ARRAY [IF (RAND()<1.0, 7, 1) , 2], ARRAY [7])", ImmutableList.of(7L));
```
Comment (cberner) [inline presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java]: add spaces, so "RAND() < 1.0"

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayIntersectFunction.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        List<Object> leftElements;
+        List<Object> rightElements;
+        try {
+            leftElements = OBJECT_MAPPER.readValue(leftArray.getInput(), COLLECTION_TYPE);
+            rightElements = OBJECT_MAPPER.readValue(rightArray.getInput(), COLLECTION_TYPE);
+        }
+        catch (IOException e) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, e);
+        }
+        leftElements.retainAll(rightElements);
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayIntersectFunction.java]: Unfortunately this doesn't work. See the developer guide on our type systems: https://prestodb.io/docs/current/develop/types.html

You need to have a custom comparison to see if two values of a type are the same. See how this works in `ArraySort`

Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayIntersectFunction.java]: Also, this is going to be O(n^2) which is too slow. You should optimize it to use a hash map, so that it runs in linear time

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayIntersectFunction.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    @Override
+    public String getDescription()
+    {
+        return "Intersection of two given arrays.";
+    }
+
+    @Override
+    public FunctionInfo specialize(Map<String, Type> types, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
+    {
+        checkArgument(types.size() == 1, format("%s expects only two arguments", FUNCTION_NAME));
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayIntersectFunction.java]: message should say "one argument"

**Comment by cberner** [comment]
Assign back to me when you've added the comments

**Comment by electrum** [comment]
Standard SQL has a set type and related functions. Is there an intersect function? If so, we should probably make this an overload of the same name.

**Comment by electrum** [comment]
This needs documentation and release notes.

**Comment by cberner** [comment]
@electrum no, it doesn't look like it. @martint and I looked at this, and there's a multiset type, but the syntax is `multiset_col1 MULTISET INTERSECT multiset_col2`

**Comment by electrum** [comment]
How does this handle nulls? We should probably do it the same way as the standard `MULTISET INTERSECT` operator.

File: presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java [hunk 4]
```diff
@@ -282,6 +282,20 @@ public void testSort()
     }
 
     @Test
+    public void testArrayIntersect()
```
Comment (electrum) [inline presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java]: This needs tests with nulls and empty array.

**Comment by cberner** [comment]
It should handle NULLs the same way that INTERSECT does. From the spec, MULTISET INTERSECT is equivalent to

```
( CASE WHEN OP1 IS NULL OR OP2 IS NULL THEN NULL
ELSE MULTISET ( SELECT T1.V
FROM UNNEST (OP1) AS T1(V)
INTERSECT SQ
SELECT T2.V
FROM UNNEST (OP2) AS T2(V)
)
END )
```

File: presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java (hunk 1 above)
Comment (ameydhar) [inline presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java]: Fixed.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayIntersectFunction.java (hunk 2 above)
Comment (ameydhar) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayIntersectFunction.java]: Implemented this function by sorting the arrays and comparing them as discussed.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayIntersectFunction.java (hunk 3 above)
Comment (ameydhar) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayIntersectFunction.java]: Fixed

**Comment by cberner** [comment]
Oh sorry, didn't notice you had updated this (there's no notification, so you have to ping me). Can you rebase this, if you'd still like to get it in?

**Comment by ameydhar** [comment]
@cberner No problem. I'll rebase and let you know when this is ready.

File: presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java [hunk 5]
```diff
@@ -328,6 +328,20 @@ public void testSort()
...
+            throws Exception
+    {
+        assertFunction("ARRAY_INTERSECT(ARRAY [12], ARRAY [10])", Lists.newArrayList());
+        assertFunction("ARRAY_INTERSECT(ARRAY['foo', 'bar', 'baz'] , ARRAY['foo', 'test', 'bar'])", ImmutableList.of("bar", "foo"));
```
Comment (cberner) [inline presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java]: add a space after "ARRAY" to match other tests in this file, and no space after ","

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayIntersectFunction.java [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+        {
+            @Override
+            public int compare(Integer r1, Integer r2)
+            {
+                return type.compareTo(rightBlock, r1, rightBlock, r2);
+            }
+        });
+
+        BlockBuilder leftBlockBuilder = new VariableWidthBlockBuilder(new BlockBuilderStatus(), leftBlock.getSizeInBytes());
+        BlockBuilder rightBlockBuilder = new VariableWidthBlockBuilder(new BlockBuilderStatus(), rightBlock.getSizeInBytes());
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayIntersectFunction.java]: Building these blocks will be a bit expensive. Instead I would fuse it with the loop below, by iterating over the leftPositions and rightPositions to get them in sorted order

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayIntersectFunction.java [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+        List<Integer> rightPositions = Ints.asList(new int[rightPositionCount]);
+
+        for (int i = 0; i < leftPositionCount; i++) {
+            leftPositions.set(i, i);
+        }
+        for (int i = 0; i < rightPositionCount; i++) {
+            rightPositions.set(i, i);
+        }
+
+        Collections.sort(leftPositions, new Comparator<Integer>()
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayIntersectFunction.java]: I think it will be more efficient if you use `IntArrays.quickSort()` because then we can avoid the boxing of `Integer`

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayIntersectFunction.java [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+    {
+        Block leftBlock = readStructuralBlock(leftArray);
+        Block rightBlock = readStructuralBlock(rightArray);
+
+        int leftPositionCount = leftBlock.getPositionCount();
+        int rightPositionCount = rightBlock.getPositionCount();
+        int leftPosition = 0;
+        int rightPosition = 0;
+
+        List<Integer> leftPositions = Ints.asList(new int[leftPositionCount]);
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayIntersectFunction.java]: Use `int[]` instead to avoid allocating individual objects for each index. You'll then need to use one of the fastutils classes for the sort, but that looks pretty simple

**Comment by cberner** [comment]
Back to you, @ameydhar , for a few comments. The approach looks right though

File: presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java (hunk 5 above)
Comment (ameydhar) [inline presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java]: Rebased and fixed these test cases

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayIntersectFunction.java (hunk 6 above)
Comment (ameydhar) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayIntersectFunction.java]: Nice suggestion. Fixed.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayIntersectFunction.java (hunk 7 above)
Comment (ameydhar) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayIntersectFunction.java]: Fixed. It took me a while to understand what 'boxing of Integer' meant :)

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayIntersectFunction.java [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+import static com.facebook.presto.util.Reflection.methodHandle;
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.lang.String.format;
+
+public final class ArrayIntersectFunction
+        extends ParametricScalar
+{
+    public static final ArrayIntersectFunction ARRAY_INTERSECT_FUNCTION = new ArrayIntersectFunction();
+    private static final String FUNCTION_NAME = "array_intersect";
+    private static final Signature SIGNATURE = new Signature(FUNCTION_NAME, ImmutableList.of(comparableTypeParameter("E")), "array<E>", ImmutableList.of("array<E>", "array<E>"), false, false);
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayIntersectFunction.java]: this actually needs to be `orderableTypeParameter` since the implementation relies on `compareTo` which is only available on orderable types

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayIntersectFunction.java [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+        return new FunctionInfo(signature, getDescription(), isHidden(), methodHandle, isDeterministic(), false, ImmutableList.of(false, false));
+    }
+
+    private static IntComparator IntBlockCompare(Type type, Block inBlock)
+    {
+        return new IntComparator() {
+            @Override
+            public int compare(int l, int r)
+            {
+                return type.compareTo(inBlock, l, inBlock, r);
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayIntersectFunction.java]: you need to handle `NULL` positions in the block

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayIntersectFunction.java [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+        IntArrays.quickSort(leftPositions, IntBlockCompare(type, leftBlock));
+        IntArrays.quickSort(rightPositions, IntBlockCompare(type, rightBlock));
+
+        BlockBuilder resultBlockBuilder = new VariableWidthBlockBuilder(new BlockBuilderStatus(), leftBlock.getSizeInBytes());
+
+        int leftCurPos = 0;
+        int rightCurPos = 0;
+
+        while (leftCurPos < leftPositionCount && rightCurPos < rightPositionCount) {
+            int compareValue = type.compareTo(leftBlock, leftPositions[leftCurPos], rightBlock, rightPositions[rightCurPos]);
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayIntersectFunction.java]: you need to handle `NULL` positions here as well

**Comment by cberner** [comment]
Ping me when you've fixed the issues with nulls

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayIntersectFunction.java [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+        int rightCurPos = 0;
+
+        while (leftCurPos < leftPositionCount && leftBlock.isNull(leftPositions[leftCurPos])) {
+            leftCurPos++;
+        }
+        while (rightCurPos < rightPositionCount && rightBlock.isNull(rightPositions[rightCurPos])) {
+            rightCurPos++;
+        }
+        while (leftCurPos < leftPositionCount && rightCurPos < rightPositionCount) {
+            int compareValue = type.compareTo(leftBlock, leftPositions[leftCurPos], rightBlock, rightPositions[rightCurPos]);
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayIntersectFunction.java]: You need to check for NULL again here

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayIntersectFunction.java [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+            if (compareValue > 0) {
+                rightCurPos++;
+            }
+            else if (compareValue < 0) {
+                leftCurPos++;
+            }
+            else {
+                type.appendTo(leftBlock, leftPositions[leftCurPos], resultBlockBuilder);
+                leftCurPos++;
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayIntersectFunction.java]: You might need to advance more than one position here. Otherwise you might output a duplicate. For example: `array_intersect(array[1, 1], array[1, 1])` should be `array[1]` not `array[1, 1]`

File: presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java [hunk 14]
```diff
@@ -328,6 +328,22 @@ public void testSort()
     }
 
     @Test
+    public void testArrayIntersect()
+            throws Exception
+    {
+        assertFunction("ARRAY_INTERSECT(ARRAY [12], ARRAY [10])", new ArrayType(BIGINT), ImmutableList.of());
+        assertFunction("ARRAY_INTERSECT(ARRAY ['foo', 'bar', 'baz'], ARRAY ['foo', 'test', 'bar'])", new ArrayType(VARCHAR), ImmutableList.of("bar", "foo"));
+        assertFunction("ARRAY_INTERSECT(ARRAY [NULL], ARRAY [NULL, NULL])", new ArrayType(UNKNOWN), ImmutableList.of());
```
Comment (cberner) [inline presto-main/src/test/java/com/facebook/presto/type/TestArrayOperators.java]: This test isn't correct. The behavior should be like the SQL INTERSECT construct, so the result of this should be `array[null]`

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayIntersectFunction.java [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+        MethodHandle methodHandle = METHOD_HANDLE.bindTo(types.get("E"));
+        Signature signature = new Signature(FUNCTION_NAME, typeSignature, typeSignature, typeSignature);
+        return new FunctionInfo(signature, getDescription(), isHidden(), methodHandle, isDeterministic(), false, ImmutableList.of(false, false));
+    }
+
+    private static IntComparator IntBlockCompare(Type type, Block inBlock)
+    {
+        return new IntComparator() {
+            @Override
+            public int compare(int l, int r)
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayIntersectFunction.java]: style comment: we try to avoid single letter variables, so name these `left` and `right`

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayIntersectFunction.java [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+
+        int leftCurPos = 0;
+        int rightCurPos = 0;
+
+        while (leftCurPos < leftPositionCount && leftBlock.isNull(leftPositions[leftCurPos])) {
+            leftCurPos++;
+        }
+        while (rightCurPos < rightPositionCount && rightBlock.isNull(rightPositions[rightCurPos])) {
+            rightCurPos++;
+        }
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayIntersectFunction.java]: output the NULL, if it was in both blocks

**Comment by cberner** [comment]
ping me when you've got the NULLs working correctly

**Comment by cberner** [comment]
Merged. Thanks!
